File: python/minimum_fuel_cost_to_report_to_the_capital.py
from collections import defaultdict
import math


class Solution:
    def minimumFuelCost(self, roads: list[list[int]], seats: int) -> int:
        """time complexity: O(n), where n is the number of roads. space complexity: O(n)"""
        graph: dict[int, list[int]] = defaultdict(list)
        for src, des in roads:
            graph[src].append(des)
            graph[des].append(src)
        result: list[0] = [0]
        self.dfs(0, -1, seats, result, graph)
        return result[0]

    def dfs(
        self,
        node: int,
        parent: int,
        seats: int,
        result: list[int],
        graph: dict[int, list[int]],
    ) -> int:
        passengers: int = 0
        for child in graph[node]:
            if child != parent:
                p: int = self.dfs(child, node, seats, result, graph)
                passengers += p
                result[0] += int(math.ceil(p / seats))
        return passengers + 1


def test_minimumFuelCost_case_1():
    # arrange
    roads: list[list[int]] = [[0, 1], [0, 2], [0, 3]]
    seats: int = 5
    expected: int = 3

    # act
    solution = Solution()
    actual = solution.minimumFuelCost(roads, seats)

    # assert
    assert expected == actual


def test_minimumFuelCost_case_2():
    # arrange
    roads: list[list[int]] = [[3, 1], [3, 2], [1, 0], [0, 4], [0, 5], [4, 6]]
    seats: int = 2
    expected: int = 7

    # act
    solution = Solution()
    actual = solution.minimumFuelCost(roads, seats)

    # assert
    assert expected == actual

File: rust/src/code/minimum_fuel_cost_to_report_to_the_capital.rs
use std::collections::HashMap;

struct Solution;

impl Solution {
    pub fn minimum_fuel_cost(roads: Vec<Vec<i32>>, seats: i32) -> i64 {
        let mut result: i64 = 0;
        let mut graph: HashMap<i32, Vec<i32>> = HashMap::new();
        for road in roads {
            graph
                .entry(road[0])
                .and_modify(|x| x.push(road[1]))
                .or_insert(vec![road[1]]);
            graph
                .entry(road[1])
                .and_modify(|x| x.push(road[0]))
                .or_insert(vec![road[0]]);
        }
        Self::dfs(0, -1, seats as i64, &mut result, &graph);
        result
    }

    fn dfs(
        node: i32,
        parent: i32,
        seats: i64,
        result: &mut i64,
        graph: &HashMap<i32, Vec<i32>>,
    ) -> i64 {
        let mut passengers: i64 = 0;
        if let Some(childs) = graph.get(&node) {
            for child in childs {
                if *child != parent {
                    let p: i64 = Self::dfs(*child, node, seats, result, graph);
                    passengers += p as i64;
                    *result += (p as f64 / seats as f64).ceil() as i64;
                }
            }
        }
        passengers + 1
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_minimum_fuel_cost_case_1() {
        // arrange
        let roads: Vec<Vec<i32>> = vec![vec![0, 1], vec![0, 2], vec![0, 3]];
        let seats: i32 = 5;
        let expected: i64 = 3;

        // act
        let actual = Solution::minimum_fuel_cost(roads, seats);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_minimum_fuel_cost_case_2() {
        // arrange
        let roads: Vec<Vec<i32>> = vec![
            vec![3, 1],
            vec![3, 2],
            vec![1, 0],
            vec![0, 4],
            vec![0, 5],
            vec![4, 6],
        ];
        let seats: i32 = 2;
        let expected: i64 = 7;

        // act
        let actual = Solution::minimum_fuel_cost(roads, seats);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_minimum_fuel_cost_case_3() {
        // arrange
        let roads: Vec<Vec<i32>> = vec![];
        let seats: i32 = 1;
        let expected: i64 = 0;

        // act
        let actual = Solution::minimum_fuel_cost(roads, seats);

        // assert
        assert_eq!(expected, actual);
    }
}
